internal/worker/caasfirewaller/appfirewaller.go

```diff
@@ -0,0 +1,258 @@
+) (network.GroupedPortRanges, error) {
+	changedPortRanges, err := w.portService.GetApplicationOpenedPortsByEndpoint(ctx, w.appUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	if lastCheckPoint.EqualTo(changedPortRanges) {
+		w.logger.Debugf(ctx, "application %q opened ports are up to date, no work to be performed")
+		return lastCheckPoint, nil
+	}
+
+	w.logger.Infof(ctx, "applying application %q updated port changes", w.appUUID)
+	err = mutator.UpdatePorts(toServicePorts(changedPortRanges), false)
+	if err != nil {
+		return nil, errors.Errorf(
+			"updating application %q ports in broker: %w", w.appUUID, err,
+		)
+	}
+
+	return changedPortRanges, nil
+}
+
+// getPortMutator returns the portMutator for the current application.
+func (w *appFirewaller) getPortMutator(ctx context.Context) (PortMutator, error) {
+	appName, err := w.applicationService.GetApplicationName(ctx, w.appUUID)
+	if err != nil {
+		return nil, errors.Errorf("getting application %q name: %w", w.appUUID, err)
+	}
+
+	app := w.broker.Application(appName, caas.DeploymentStateful)
+	return app, nil
+}
+
+// Kill is part of the worker.Worker interface.
+// Kill sets this [appFirewaller] to dying with a nil error shutting down the
+// main loop and any child workers.
+//
+// Implements the [worker.Worker] interface.
+func (w *appFirewaller) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// loop is the main processing routine of the worker waiting for applications to
+// be added or removed from the model. For each application in the model a child
+// worker will be created to manage the open ports of the application.
+//
+// loop returns when the worker is placed into a dying state or an expected
+// error occurs.
+func (w *appFirewaller) loop() (err error) {
+	ctx := w.catacomb.Context(context.Background())
+	defer func() {
+		// If the application has been deleted, we can return nil.
+		// If are returning because of an application not found error then
+		// return a nil error as this worker has nothing more to do. The
+		// application has gone away. Returning will force the catacomb to shut
+		// down the watchers used.
+		if errors.Is(err, domainapplicationerrors.ApplicationNotFound) {
+			w.logger.Debugf(
+				ctx,
+				"application %q removed, shutting down caas firewaller",
+				w.appUUID,
+			)
+			err = nil
+		}
+	}()
+
+	portMutator, err := w.getPortMutator(ctx)
+	if err != nil {
+		return errors.Errorf(
+			"getting application %q port mutator: %w",
+			w.appUUID, err,
+		)
+	}
+
+	// lastCheckPoint keeps track of the last known port ranges that have been
+	// applied to the application.
+	lastCheckPoint := network.GroupedPortRanges{}
+
+	// Ensure the applications initial set of open port requirements are applied.
+	lastCheckPoint, err = w.ensureOpenPorts(ctx, portMutator, lastCheckPoint)
+	if err != nil {
+		return errors.Errorf(
+			"ensuring initial open port requirements for application %q are applied: %w",
+			w.appUUID, err,
+		)
+	}
```

Comment:
`WatchOpenedPortsForApplication` must fire an initial event, if it does not, then it is incorrect. Since it must fire an initial event, then it is OK to remove this block.